<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Описание-проекта" data-toc-modified-id="Описание-проекта-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Описание проекта</a></span><ul class="toc-item"><li><span><a href="#Описание-данных" data-toc-modified-id="Описание-данных-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Описание данных</a></span></li></ul></li><li><span><a href="#Подключение-к-БД,-исследование-таблиц" data-toc-modified-id="Подключение-к-БД,-исследование-таблиц-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Подключение к БД, исследование таблиц</a></span></li><li><span><a href="#Задачи" data-toc-modified-id="Задачи-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Задачи</a></span><ul class="toc-item"><li><span><a href="#Посчитайте,-сколько-книг-вышло-после-1-января-2000-года" data-toc-modified-id="Посчитайте,-сколько-книг-вышло-после-1-января-2000-года-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Посчитайте, сколько книг вышло после 1 января 2000 года</a></span></li><li><span><a href="#Для-каждой-книги-посчитайте-количество-обзоров-и-среднюю-оценку" data-toc-modified-id="Для-каждой-книги-посчитайте-количество-обзоров-и-среднюю-оценку-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Для каждой книги посчитайте количество обзоров и среднюю оценку</a></span></li><li><span><a href="#Определите-издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц-—-так-вы-исключите-из-анализа-брошюры" data-toc-modified-id="Определите-издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц-—-так-вы-исключите-из-анализа-брошюры-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры</a></span></li><li><span><a href="#Определите-автора-с-самой-высокой-средней-оценкой-книг-—-учитывайте-только-книги-с-50-и-более-оценками" data-toc-modified-id="Определите-автора-с-самой-высокой-средней-оценкой-книг-—-учитывайте-только-книги-с-50-и-более-оценками-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками</a></span></li><li><span><a href="#Посчитайте-среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-50-оценок" data-toc-modified-id="Посчитайте-среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-50-оценок-3.5"><span class="toc-item-num">3.5&nbsp;&nbsp;</span>Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок</a></span></li></ul></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Вывод</a></span></li></ul></div>

# Анализ БД сервиса для чтения книг

## Описание проекта

Компания приобрела крупный сервис для чтения книг по подписке. Этот сервис имеет базу данных. В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг.\
Задача проекта — проанализировать базу данных, чтобы получить возможность сформулировать ценностное предложение для нового продукта.

### Описание данных

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

## Подключение к БД, исследование таблиц

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine

In [2]:
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
            'pwd': 'Sdf4$2;d-d30pp', # пароль
            'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
            'port': 6432, # порт подключения
            'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                 db_config['pwd'],
                                                 db_config['host'],
                                                 db_config['port'],
                                                 db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

Изучим таблицы базы данных.

**books**

Первые строки.

In [3]:
query = '''
            SELECT * 
            FROM books
            LIMIT 5
        '''

In [4]:
pd.io.sql.read_sql(query, con = engine) 

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

Количество записей.

In [5]:
query = '''
            SELECT COUNT(*)
            FROM books
            
        '''

In [6]:
pd.io.sql.read_sql(query, con = engine) 

count
0   1000

Первые строки.

In [7]:
query = '''
            SELECT * 
            FROM authors
            LIMIT 5
        '''

In [8]:
pd.io.sql.read_sql(query, con = engine) 

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

Количество записей.

In [9]:
query = '''
            SELECT COUNT(*)
            FROM authors
            
        '''

In [10]:
pd.io.sql.read_sql(query, con = engine) 

count
0    636

Первые строки.

In [11]:
query = '''
            SELECT * 
            FROM publishers
            LIMIT 5
        '''

In [12]:
pd.io.sql.read_sql(query, con = engine) 

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

Количество записей.

In [13]:
query = '''
            SELECT COUNT(*)
            FROM publishers
            
        '''

In [14]:
pd.io.sql.read_sql(query, con = engine) 

count
0    340

Первые строки.

In [15]:
query = '''
            SELECT * 
            FROM ratings
            LIMIT 5
        '''

In [16]:
pd.io.sql.read_sql(query, con = engine) 

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

Количество записей.

In [17]:
query = '''
            SELECT COUNT(*)
            FROM ratings
            
        '''

In [18]:
pd.io.sql.read_sql(query, con = engine) 

count
0   6456

Количество книг, у которых есть оценки.

In [19]:
query = '''
            SELECT COUNT(DISTINCT book_id)
            FROM ratings
            
        '''

In [20]:
pd.io.sql.read_sql(query, con = engine) 

count
0   1000

Первые строки.

In [21]:
query = '''
            SELECT * 
            FROM reviews
            LIMIT 5
        '''

In [22]:
pd.io.sql.read_sql(query, con = engine) 

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

Количество записей.

In [23]:
query = '''
            SELECT COUNT(*)
            FROM reviews
            
        '''

In [24]:
pd.io.sql.read_sql(query, con = engine) 

count
0   2793

Количество книг, на которые есть отзывы.

In [25]:
query = '''
            SELECT COUNT(DISTINCT book_id)
            FROM reviews
            
        '''

In [26]:
pd.io.sql.read_sql(query, con = engine) 

count
0    994

Всего в базе данных хранится информация о 1000 книг 636 авторов, выпущенных в 340 издательствах. У каждой книги есть оценки, всего оценок 6456. На 994 книги имеется 2793 отзыва.

## Задачи

### Посчитайте, сколько книг вышло после 1 января 2000 года

In [27]:
query = '''
            SELECT COUNT(*) 
            FROM books
            WHERE publication_date::DATE > '2000-01-01'
        '''

In [28]:
pd.io.sql.read_sql(query, con = engine)

count
0    819

Большая часть книг в базе данных относительно новые - 819 из 1000 вышли в 2000 году и позднее.

### Для каждой книги посчитайте количество обзоров и среднюю оценку

In [29]:
query = '''
            WITH rev AS
            (
            SELECT DISTINCT book_id,
                    COUNT(review_id) OVER (PARTITION BY book_id) AS review_count
            FROM reviews
            ),
            
            rat AS
            (
            SELECT DISTINCT book_id,
                    AVG(rating) OVER (PARTITION BY book_id) AS avg_rating
            FROM ratings
            )
            
            SELECT rat.book_id,
                    books.title,
                    review_count,
                    avg_rating
            FROM rev 
            RIGHT JOIN rat ON rev.book_id=rat.book_id
            JOIN books ON rat.book_id = books.book_id
            
        '''

In [30]:
pd.io.sql.read_sql(query, con = engine)

book_id                                              title  review_count  \
0          1                                       'Salem's Lot           2.0   
1          2                 1 000 Places to See Before You Die           1.0   
2          3  13 Little Blue Envelopes (Little Blue Envelope...           3.0   
3          4  1491: New Revelations of the Americas Before C...           2.0   
4          5                                               1776           4.0   
..       ...                                                ...           ...   
995      996           Wyrd Sisters (Discworld  #6; Witches #2)           3.0   
996      997                        Xenocide (Ender's Saga  #3)           3.0   
997      998                                    Year of Wonders           4.0   
998      999                        You Suck (A Love Story  #2)           2.0   
999     1000  Zen and the Art of Motorcycle Maintenance: An ...           4.0   

     avg_rating  
0      3.666667  
1      2.500000  
2      4.666667  
3      4.500000  
4      4.000000  
..          ...  
995    3.666667  
996    3.400000  
997    3.200000  
998    4.500000  
999    3.833333  

[1000 rows x 4 columns]

В полученной таблице на каждую из 1000 книг указано количество отзывов и средняя оценка.

### Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры

In [31]:
query = '''
            WITH pub AS 
            (SELECT DISTINCT publisher_id,
                    COUNT(book_id) OVER (PARTITION BY publisher_id) AS book_count
            FROM books
            WHERE num_pages > 50
            ORDER BY book_count DESC
            LIMIT 1)
            
            SELECT pub.publisher_id,
                    publisher
            FROM publishers
            RIGHT JOIN pub ON publishers.publisher_id=pub.publisher_id
        '''

In [32]:
pd.io.sql.read_sql(query, con = engine)

publisher_id      publisher
0           212  Penguin Books

Издательство "Penguin Books" выпустило больше всего книг.

### Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками

In [33]:
query = '''
            WITH count_r AS
            (SELECT book_id
            FROM (SELECT DISTINCT book_id,
                        COUNT(rating) OVER (PARTITION BY book_id) AS count_rating
                 FROM ratings) AS c_r
            WHERE count_rating >= 50),
            
            id AS
            (SELECT DISTINCT count_r.book_id,
                    AVG(rating) OVER (PARTITION BY count_r.book_id) AS avg_rating
            FROM ratings
            RIGHT JOIN count_r ON ratings.book_id = count_r.book_id
            ORDER BY avg_rating DESC
            LIMIT 1)
            
            SELECT authors.author_id,
                    author
            FROM authors
            JOIN books ON authors.author_id = books.author_id
            RIGHT JOIN id ON books.book_id = id.book_id
            
        '''

In [34]:
pd.io.sql.read_sql(query, con = engine) 

author_id                      author
0        236  J.K. Rowling/Mary GrandPré

Самый популярный автор с максимальной средней оценкой - J.K. Rowling/Mary GrandPré.

### Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок

In [35]:
query = '''
            WITH uuser AS
            (SELECT username
            FROM (SELECT DISTINCT username,
                        COUNT(rating) OVER (PARTITION BY username) AS count_rating
                 FROM ratings) AS c_r
            WHERE count_rating >= 50),
            
            rev AS
            (SELECT DISTINCT uuser.username,
                    COUNT(review_id) OVER (PARTITION BY uuser.username) AS count_review
            FROM uuser
            LEFT JOIN reviews ON uuser.username = reviews.username)
            
            SELECT AVG(count_review)
            FROM rev
        '''

In [36]:
pd.io.sql.read_sql(query, con = engine) 

avg
0  24.222222

Пользователи, которые поставили больше всего оценок, написали в среднем по 24 отзыва.

## Вывод

В ходе проекта проанализирована база данных недавно приобретенного книжного сервиса.
- Всего в базе содержится информация о 1000 книг 636 авторов, выпущенных в 340 издательствах. У каждой книги есть оценки, всего оценок 6456. На 994 книги имеется 2793 отзыва.
- 82% книг в базе данных вышли в 2000 году и позднее.
- Составлена таблица со средним рейтингом и количеством отзывов на каждую книгу.
- Больше всего книг в сервисе от издательства "Penguin Books".
- Самый популярный автор в сервисе - J.K. Rowling/Mary GrandPré.
- Самые активные пользователи сервиса поставили более 50 оценок и написали в среднем по 24 отзыва.

Проведенный анализ дает возможность сформировать ценностное предложение для нового продукта.